### PR TITLE
HAVE_PTHREAD gate in test.h

### PR DESCRIPTION
**Pull request by gojimmypi**
# Description

This is a corner case of cross-compiling testwolfcrypt, in my case for the Xtensa ESP32-S3 Linux kernel.

I didn't have a `pthread.h` and noticed that several other places (but not all?) are gated with `HAVE_PTHREAD`. 

This PR adds the `HAVE_PTHREAD` only to  `test.h` that I encountered, but I can add to other places in this PR if desired

Fixes zd# n/a/

# Testing

How did you test?

Ran the test app in cross-compiled app as well as Linux & ESP-IDF for ESP32.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Review by dgarske — APPROVED** [review]

**Review by julek-wolfssl — APPROVED** [review]

File: wolfssl/test.h [hunk 1]
```diff
@@ -182,7 +182,9 @@
     #include <sys/ioctl.h>
     #include <sys/time.h>
     #include <sys/socket.h>
-    #include <pthread.h>
+    #ifdef HAVE_PTHREAD
```
Comment (dgarske) [inline wolfssl/test.h:185]: Some Jenkins jobs are failing with this. The `HAVE_PTHREAD` comes from `config.h`. Please add this at top of test.h:

```
#ifdef HAVE_CONFIG_H
    #include <config.h>
#endif
```

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfssl/test.h (hunk 1 above)
Comment (gojimmypi) [reply wolfssl/test.h:189]: @dgarske we were talking about two different `test.h` files, one in [wolfssl/test.h](https://github.com/wolfSSL/wolfssl/blob/master/wolfssl/test.h) and the other in [wolfcrypt/test/test.h](https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/test/test.h). 

I applied your suggested change to both, but the PRB is still failing.

The issue is a bit more complex in that `test.h` [uses](https://github.com/wolfSSL/wolfssl/blob/master/wolfssl/test.h#L2204) `_POSIX_THREADS` not `HAVE_PTHREAD`. There are of course a variety of ways and places to handle that.

What do you think of adding something like this to the end of [wolfssl/wolfcrypt/settings.h](https://github.com/wolfSSL/wolfssl/blob/2f44ece4d840354aed4a391ec408d01b706bbe57/wolfssl/wolfcrypt/settings.h#L3099) ?

```
#ifdef _POSIX_THREADS
    #define HAVE_PTHREAD 
#endif
```

Other suggestions?

**Comment by gojimmypi** [comment]
After seeing the PRB failures, it seems there was some logic missing from a few of the examples. In particular, when the threading capability _exists_ but is explicitly _turned off_ with `--enable-singlethreaded` to force single-threaded mode. 

I sprinkled some `&& !defined(SINGLE_THREADED)` logic in places where `if defined(_POSIX_THREADS)` is used.  It appears this logic was already in use in other places.

There may be an argument to combine all that logic into a single global definition.

File: wolfssl/wolfcrypt/settings.h [hunk 2]
```diff
@@ -52,6 +52,12 @@
     extern "C" {
 #endif
 
+/* pick up compiler def; may need to turn on HAVE_PTHREAD for some configs */
+#ifdef _POSIX_THREADS
+    #undef HAVE_PTHREAD
```
Comment (dgarske) [inline wolfssl/wolfcrypt/settings.h:57]: This looks good to me. Now that this is in settings.h I think it best to change the `_POSIX_THREADS` elsewhere to use `HAVE_PTHREAD`. I also think you should add ` && !defined(SINGLE_THREADED)` and remove it from the refactor you did.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: wolfssl/wolfcrypt/settings.h (hunk 2 above)
Comment (gojimmypi) [reply wolfssl/wolfcrypt/settings.h]: cool. I've applied your suggested changes and changed the PR description & title to reflect the global nature of this change

**Comment by gojimmypi** [comment]
@dgarske  - After completing the desired changes for `HAVE_PTHREAD` and  `_POSIX_THREADS`, I noticed there's a potentially undesired `WOLFSSL_PTHREADS` that should also be refactored? See the [codebase for references](https://github.com/search?q=repo%3AwolfSSL%2Fwolfssl%20WOLFSSL_PTHREADS&type=code).

If you agree, changes to (or more specifically: removal of) `WOLFSSL_PTHREADS` would probably be in a separate PR & I'd be happy to work on that right away.

**Review by dgarske — APPROVED** [review]
Very nice cleanup!

**Comment by gojimmypi** [comment]
See also https://github.com/wolfSSL/wolfssl/pull/6545
